smithyctl/internal/images/component.go

```diff
@@ -0,0 +1,193 @@
+package images
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"regexp"
+	"slices"
+	"strings"
+
+	"github.com/go-errors/errors"
+	"github.com/google/go-containerregistry/pkg/name"
+
+	v1 "github.com/smithy-security/smithy/pkg/types/v1"
+)
+
+var (
+	pluralToSingularComponentType = map[string]string{}
+	componentPluralisedTypeChoice = strings.Join(
+		slices.Collect(
+			func(yield func(string) bool) {
+				for _, ct := range v1.ComponentTypeNames() {
+					if ct != v1.ComponentTypeUnknown.String() {
+						pluralToSingularComponentType[ct+"s"] = ct
+						if !yield(ct + "s") {
+							return
+						}
+					}
+				}
+			},
+		),
+		"|",
+	)
+	componentRepositoryRegex = regexp.MustCompile(fmt.Sprintf("([a-z-_]+/)?(%s)/[a-z-_]+", componentPluralisedTypeChoice))
+	// ErrNotAComponentRepo is returns when an image repository doesn't match
+	// the expected format
+	ErrNotAComponentRepo = errors.Errorf("image repository doesn't match the expected component path: <component root directory>/%s/component.yaml", componentPluralisedTypeChoice)
+)
+
+// resolutionOptions is a struct that defines common properties of all the
+// images of components managed by the system
+type resolutionOptions struct {
+	registry  string
+	namespace string
+	tag       string
+}
+
+// ResolutionOptionFn is used to define common attributes of all the images
+// of components
+type ResolutionOptionFn func(*resolutionOptions)
```

Comment:
Nit but it would be nice to return an erro if the passed arg is misconfigured :)

So `type ResolutionOptionFn func(*resolutionOptions) error`